File: frontend/src/routes/_shared/protocol.ts
import { requestJson } from './api';

export type NormalizedValueItem = {
  value?: number | null;
  unit?: string | null;
  raw_value?: string | null;
  operator?: string | null;
  min_value?: number | null;
  max_value?: number | null;
  status?: string | null;
};

export type ConditionItem = {
  temperature?: NormalizedValueItem | null;
  duration?: NormalizedValueItem | null;
  pressure?: NormalizedValueItem | null;
  heating_rate?: NormalizedValueItem | null;
  cooling_rate?: NormalizedValueItem | null;
  ph?: NormalizedValueItem | null;
  atmosphere?: string | null;
  environment?: string | null;
  raw_text?: string | null;
};

export type MaterialRefItem = {
  name?: string | null;
  formula?: string | null;
  role?: string | null;
  amount?: string | null;
  composition_note?: string | null;
  grade?: string | null;
  source_text?: string | null;
};

export type MeasurementSpecItem = {
  method?: string | null;
  instrument?: string | null;
  target_property?: string | null;
  metrics?: string[] | null;
  conditions?: Record<string, unknown> | null;
  output_ref?: string | null;
  source_text?: string | null;
};

export type ControlSpecItem = {
  control_type?: string | null;
  description?: string | null;
  rationale?: string | null;
  source_text?: string | null;
};

export type EvidenceRefItem = {
  paper_id?: string | null;
  section_id?: string | null;
  block_id?: string | null;
  snippet_id?: string | null;
  section_type?: string | null;
  page_start?: number | null;
  page_end?: number | null;
  figure_or_table?: string | null;
  quote_span?: string | null;
  source_text?: string | null;
  confidence_score?: number | null;
};

export type ProtocolStepItem = {
  step_id: string;
  paper_id: string;
  paper_title?: string | null;
  order?: number | null;
  action: string;
  section_id?: string | null;
  block_id?: string | null;
  phase?: string | null;
  materials: MaterialRefItem[];
  conditions?: ConditionItem | null;
  purpose?: string | null;
  expected_output?: string | null;
  characterization: MeasurementSpecItem[];
  controls: ControlSpecItem[];
  evidence_refs: EvidenceRefItem[];
  confidence_score?: number | null;
};

export type ProtocolStepListResponse = {
  collection_id?: string;
  output_path?: string;
  count: number;
  items: ProtocolStepItem[];
};

export type ProtocolSearchHit = {
  step_id: string;
  paper_id: string;
  paper_title?: string | null;
  section_id?: string | null;
  block_id?: string | null;
  action: string;
  matched_fields: string[];
  excerpt?: string | null;
  score?: number | null;
};

export type ProtocolSearchResponse = {
  collection_id?: string;
  query: string;
  output_path?: string;
  count: number;
  items: ProtocolSearchHit[];
};

export type SOPDraftItem = {
  sop_id?: string | null;
  objective?: string | null;
  hypothesis?: string | null;
  variables?: Record<string, unknown> | null;
  constraints?: Record<string, unknown> | null;
  controls: ControlSpecItem[];
  steps: ProtocolStepItem[];
  measurement_plan: MeasurementSpecItem[];
  acceptance_criteria: string[];
  risks: string[];
  open_questions: string[];
  review_status?: string | null;
};

export type SOPDraftResponse = {
  collection_id?: string;
  output_path?: string;
  sop_draft: SOPDraftItem;
  warnings: string[];
};

function normalizeStep(item: unknown): ProtocolStepItem | null {
  if (!item || typeof item !== 'object') return null;
  const record = item as Record<string, unknown>;
  const stepId = String(record.step_id ?? '').trim();
  const paperId = String(record.paper_id ?? '').trim();
  const action = String(record.action ?? '').trim();
  if (!stepId || !paperId || !action) return null;

  return {
    step_id: stepId,
    paper_id: paperId,
    paper_title: typeof record.paper_title === 'string' ? record.paper_title : null,
    order: typeof record.order === 'number' ? record.order : Number(record.order ?? 0),
    action,
    section_id: typeof record.section_id === 'string' ? record.section_id : null,
    block_id: typeof record.block_id === 'string' ? record.block_id : null,
    phase: typeof record.phase === 'string' ? record.phase : null,
    materials: Array.isArray(record.materials) ? (record.materials as MaterialRefItem[]) : [],
    conditions: record.conditions && typeof record.conditions === 'object' ? (record.conditions as ConditionItem) : null,
    purpose: typeof record.purpose === 'string' ? record.purpose : null,
    expected_output: typeof record.expected_output === 'string' ? record.expected_output : null,
    characterization: Array.isArray(record.characterization)
      ? (record.characterization as MeasurementSpecItem[])
      : [],
    controls: Array.isArray(record.controls) ? (record.controls as ControlSpecItem[]) : [],
    evidence_refs: Array.isArray(record.evidence_refs) ? (record.evidence_refs as EvidenceRefItem[]) : [],
    confidence_score:
      typeof record.confidence_score === 'number'
        ? record.confidence_score
        : Number(record.confidence_score ?? 0)
  };
}

export async function listProtocolSteps(
  collectionId: string,
  options: { paperId?: string; blockType?: string; limit?: number; offset?: number } = {}
) {
  const params = new URLSearchParams();
  if (options.paperId?.trim()) params.set('paper_id', options.paperId.trim());
  if (options.blockType?.trim()) params.set('block_type', options.blockType.trim());
  params.set('limit', String(options.limit ?? 20));
  params.set('offset', String(options.offset ?? 0));

  const data = (await requestJson(
    `/collections/${encodeURIComponent(collectionId)}/protocol/steps?${params.toString()}`,
    { method: 'GET' }
  )) as Record<string, unknown>;

  const items = Array.isArray(data.items)
    ? data.items.map((item) => normalizeStep(item)).filter((item): item is ProtocolStepItem => item !== null)
    : [];

  return {
    collection_id: typeof data.collection_id === 'string' ? data.collection_id : collectionId,
    output_path: typeof data.output_path === 'string' ? data.output_path : undefined,
    count: typeof data.count === 'number' ? data.count : items.length,
    items
  } satisfies ProtocolStepListResponse;
}

export async function searchProtocolSteps(
  collectionId: string,
  options: { query: string; paperId?: string; limit?: number }
) {
  const params = new URLSearchParams({ q: options.query.trim(), limit: String(options.limit ?? 10) });
  if (options.paperId?.trim()) params.set('paper_id', options.paperId.trim());

  const data = (await requestJson(
    `/collections/${encodeURIComponent(collectionId)}/protocol/search?${params.toString()}`,
    { method: 'GET' }
  )) as Record<string, unknown>;

  return {
    collection_id: typeof data.collection_id === 'string' ? data.collection_id : collectionId,
    query: String(data.query ?? options.query),
    output_path: typeof data.output_path === 'string' ? data.output_path : undefined,
    count: typeof data.count === 'number' ? data.count : 0,
    items: Array.isArray(data.items)
      ? data.items.map((item) => {
          const record = item as Record<string, unknown>;
          return {
            step_id: String(record.step_id ?? ''),
            paper_id: String(record.paper_id ?? ''),
            paper_title: typeof record.paper_title === 'string' ? record.paper_title : null,
            section_id: typeof record.section_id === 'string' ? record.section_id : null,
            block_id: typeof record.block_id === 'string' ? record.block_id : null,
            action: String(record.action ?? ''),
            matched_fields: Array.isArray(record.matched_fields)
              ? record.matched_fields.map((value) => String(value))
              : Array.isArray(record.matched_terms)
                ? record.matched_terms.map((value) => String(value))
                : [],
            excerpt: typeof record.excerpt === 'string' ? record.excerpt : null,
            score:
              typeof record.score === 'number' ? record.score : Number.isFinite(Number(record.score)) ? Number(record.score) : null
          } satisfies ProtocolSearchHit;
        })
      : []
  } satisfies ProtocolSearchResponse;
}

export async function generateProtocolSop(
  collectionId: string,
  payload: {
    goal: string;
    targetProperties?: string[];
    paperIds?: string[];
    maxSteps?: number;
  }
) {
  const data = (await requestJson(`/collections/${encodeURIComponent(collectionId)}/protocol/sop`, {
    method: 'POST',
    body: JSON.stringify({
      goal: payload.goal,
      target_properties: payload.targetProperties ?? [],
      paper_ids: payload.paperIds ?? [],
      max_steps: payload.maxSteps ?? 8
    })
  })) as Record<string, unknown>;

  return {
    collection_id: typeof data.collection_id === 'string' ? data.collection_id : collectionId,
    output_path: typeof data.output_path === 'string' ? data.output_path : undefined,
    sop_draft: (data.sop_draft as SOPDraftItem | undefined) ?? {
      controls: [],
      steps: [],
      measurement_plan: [],
      acceptance_criteria: [],
      risks: [],
      open_questions: []
    },
    warnings: Array.isArray(data.warnings) ? data.warnings.map((item) => String(item)) : []
  } satisfies SOPDraftResponse;
}

File: backend/controllers/schemas/derived/protocol.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class NormalizedValueItem(BaseModel):
    """Normalized numeric value with raw text traceability."""

    value: float | None = Field(default=None, description="归一化后的数值")
    unit: str | None = Field(default=None, description="归一化单位，如 K、s、MPa")
    raw_value: str | None = Field(default=None, description="原始文本值")
    operator: str | None = Field(
        default=None, description="比较符，如 =、>、<、~、range"
    )
    min_value: float | None = Field(default=None, description="区间下界")
    max_value: float | None = Field(default=None, description="区间上界")
    status: Literal["reported", "inferred", "not_reported", "ambiguous"] = Field(
        default="not_reported", description="值状态"
    )


class ConditionItem(BaseModel):
    """Normalized experimental conditions for a protocol step."""

    temperature: NormalizedValueItem | None = Field(
        default=None, description="温度，统一建议为 K"
    )
    duration: NormalizedValueItem | None = Field(
        default=None, description="时长，统一建议为 s"
    )
    pressure: NormalizedValueItem | None = Field(
        default=None, description="压力，统一建议为 Pa"
    )
    heating_rate: NormalizedValueItem | None = Field(
        default=None, description="升温速率"
    )
    cooling_rate: NormalizedValueItem | None = Field(
        default=None, description="降温速率"
    )
    ph: NormalizedValueItem | None = Field(default=None, description="pH")
    atmosphere: str | None = Field(default=None, description="气氛，如 Ar、Air")
    environment: str | None = Field(default=None, description="环境补充说明")
    raw_text: str | None = Field(default=None, description="原始条件片段")


class MaterialRefItem(BaseModel):
    """Material or reagent reference used in a protocol step."""

    name: str = Field(..., description="材料或试剂名称")
    formula: str | None = Field(default=None, description="化学式")
    role: Literal[
        "precursor",
        "solvent",
        "additive",
        "matrix",
        "filler",
        "sample",
        "product",
        "other",
    ] = Field(default="other", description="材料角色")
    amount: NormalizedValueItem | None = Field(default=None, description="用量")
    composition_note: str | None = Field(default=None, description="配比或组成备注")
    grade: str | None = Field(default=None, description="牌号/等级")
    source_text: str | None = Field(default=None, description="原始片段")


class MeasurementSpecItem(BaseModel):
    """Measurement or characterization plan."""

    method: str = Field(..., description="表征或测试方法")
    instrument: str | None = Field(default=None, description="设备名称")
    target_property: str | None = Field(default=None, description="目标性质")
    metrics: list[str] = Field(default_factory=list, description="指标字段")
    conditions: dict[str, Any] = Field(
        default_factory=dict, description="测试条件补充"
    )
    output_ref: str | None = Field(default=None, description="输出或图表引用")
    source_text: str | None = Field(default=None, description="原始片段")


class ControlSpecItem(BaseModel):
    """Control or baseline definition."""

    control_type: Literal[
        "baseline",
        "blank",
        "untreated",
        "literature",
        "ablation",
        "other",
    ] = Field(default="other", description="对照类型")
    description: str = Field(..., description="对照描述")
    rationale: str | None = Field(default=None, description="对照理由")
    source_text: str | None = Field(default=None, description="原始片段")


class EvidenceRefItem(BaseModel):
    """Evidence anchor back to source paper text."""

    paper_id: str = Field(..., description="论文 ID")
    section_id: str | None = Field(default=None, description="section ID")
    block_id: str | None = Field(default=None, description="procedure block ID")
    snippet_id: str | None = Field(default=None, description="文本片段 ID")
    section_type: str | None = Field(default=None, description="section 类型")
    page_start: int | None = Field(default=None, description="起始页")
    page_end: int | None = Field(default=None, description="结束页")
    figure_or_table: str | None = Field(default=None, description="图表引用")
    quote_span: str | None = Field(default=None, description="原文摘录")
    source_text: str | None = Field(default=None, description="证据全文片段")
    confidence_score: float | None = Field(default=None, description="证据置信度")


class ProtocolStepItem(BaseModel):
    """Structured protocol step item."""

    step_id: str = Field(..., description="步骤 ID")
    paper_id: str = Field(..., description="来源论文 ID")
    paper_title: str | None = Field(default=None, description="来源论文标题")
    order: int = Field(..., description="步骤顺序")
    action: str = Field(..., description="核心动作")
    section_id: str | None = Field(default=None, description="section ID")
    block_id: str | None = Field(default=None, description="block ID")
    phase: Literal[
        "preparation",
        "synthesis",
        "post_treatment",
        "characterization",
        "property_test",
        "analysis",
        "other",
    ] = Field(default="other", description="步骤阶段")
    materials: list[MaterialRefItem] = Field(
        default_factory=list, description="材料与试剂"
    )
    conditions: ConditionItem = Field(
        default_factory=ConditionItem, description="实验条件"
    )
    purpose: str | None = Field(default=None, description="步骤目的")
    expected_output: str | None = Field(default=None, description="期望输出")
    characterization: list[MeasurementSpecItem] = Field(
        default_factory=list, description="表征与测试"
    )
    controls: list[ControlSpecItem] = Field(
        default_factory=list, description="对照信息"
    )
    evidence_refs: list[EvidenceRefItem] = Field(
        default_factory=list, description="证据锚点"
    )
    confidence_score: float | None = Field(default=None, description="步骤置信度")


class SOPDraftItem(BaseModel):
    """Structured SOP draft composed from protocol steps."""

    sop_id: str = Field(..., description="SOP ID")
    objective: str = Field(..., description="实验目标")
    hypothesis: str | None = Field(default=None, description="实验假设")
    variables: list[str] = Field(default_factory=list, description="关键变量")
    constraints: dict[str, Any] = Field(
        default_factory=dict, description="设备/温度/时间等约束"
    )
    controls: list[ControlSpecItem] = Field(
        default_factory=list, description="对照设计"
    )
    steps: list[ProtocolStepItem] = Field(
        default_factory=list, description="结构化步骤"
    )
    measurement_plan: list[MeasurementSpecItem] = Field(
        default_factory=list, description="测量与表征计划"
    )
    acceptance_criteria: list[str] = Field(
        default_factory=list, description="验收标准"
    )
    risks: list[str] = Field(default_factory=list, description="风险项")
    open_questions: list[str] = Field(default_factory=list, description="未决问题")
    review_status: Literal[
        "draft",
        "human_review_required",
        "approved",
        "changes_requested",
        "rejected",
    ] = Field(default="draft", description="审核状态")


class ProtocolExtractRequest(BaseModel):
    """Request payload to generate structured protocol artifacts."""

    output_path: str | None = Field(
        default=None, description="集合输出目录；为空时走默认配置"
    )
    paper_ids: list[str] = Field(
        default_factory=list, description="限定论文 ID 范围"
    )
    overwrite: bool = Field(default=False, description="是否覆盖已有协议产物")


class ProtocolArtifactCounts(BaseModel):
    """Counts for generated protocol artifacts."""

    sections: int = 0
    procedure_blocks: int = 0
    protocol_steps: int = 0


class ProtocolExtractResponse(BaseModel):
    """Response after protocol artifact extraction."""

    status: str = Field(..., description="执行状态")
    output_path: str = Field(..., description="协议产物目录")
    artifacts: list[str] = Field(default_factory=list, description="产物文件名")
    counts: ProtocolArtifactCounts = Field(
        default_factory=ProtocolArtifactCounts, description="产物计数"
    )
    warnings: list[str] = Field(default_factory=list, description="警告信息")


class ProtocolStepListResponse(BaseModel):
    """Response for listing extracted protocol steps."""

    output_path: str = Field(..., description="协议产物目录")
    count: int = Field(..., description="返回数量")
    items: list[ProtocolStepItem] = Field(default_factory=list, description="步骤列表")


class ProtocolSearchHit(BaseModel):
    """Single protocol search result."""

    step_id: str = Field(..., description="步骤 ID")
    paper_id: str = Field(..., description="来源论文 ID")
    paper_title: str | None = Field(default=None, description="来源论文标题")
    section_id: str | None = Field(default=None, description="section ID")
    block_id: str | None = Field(default=None, description="block ID")
    action: str = Field(..., description="命中动作")
    matched_fields: list[str] = Field(
        default_factory=list, description="命中的字段"
    )
    excerpt: str | None = Field(default=None, description="命中摘录")
    score: float | None = Field(default=None, description="匹配分数")


class ProtocolSearchResponse(BaseModel):
    """Response for protocol-level search."""

    query: str = Field(..., description="检索词")
    output_path: str = Field(..., description="协议产物目录")
    count: int = Field(..., description="返回数量")
    items: list[ProtocolSearchHit] = Field(default_factory=list, description="结果列表")


class SOPDraftRequest(BaseModel):
    """Request payload to generate an SOP draft."""

    output_path: str | None = Field(
        default=None, description="集合输出目录；为空时走默认配置"
    )
    goal: str = Field(..., description="实验目标")
    paper_ids: list[str] = Field(default_factory=list, description="限定论文 ID")
    step_ids: list[str] = Field(default_factory=list, description="限定步骤 ID")
    target_properties: list[str] = Field(
        default_factory=list, description="目标性质"
    )
    constraints: dict[str, Any] = Field(
        default_factory=dict, description="实验约束"
    )
    max_steps: int | None = Field(default=None, description="最大步骤数")


class SOPDraftResponse(BaseModel):
    """Response containing a structured SOP draft."""

    output_path: str = Field(..., description="协议产物目录")
    sop_draft: SOPDraftItem = Field(..., description="结构化 SOP 草案")
    warnings: list[str] = Field(default_factory=list, description="警告信息")
